dids/diddht/diddht.go

```diff
@@ -1,175 +1,191 @@
 package diddht
 
 import (
+	"context"
 	"errors"
 	"fmt"
-	"strings"
-
-	"golang.org/x/net/dns/dnsmessage"
+	"net/http"
+	"sync"
+	"time"
 
+	"github.com/tbd54566975/web5-go/crypto"
+	"github.com/tbd54566975/web5-go/crypto/dsa"
+	"github.com/tbd54566975/web5-go/dids/did"
 	"github.com/tbd54566975/web5-go/dids/didcore"
+	"github.com/tbd54566975/web5-go/dids/diddht/internal/bep44"
+	"github.com/tbd54566975/web5-go/dids/diddht/internal/dns"
+	"github.com/tbd54566975/web5-go/dids/diddht/internal/pkarr"
+	"github.com/tv42/zbase32"
 )
 
-// Decoder is used to structure the DNS representation of a DID
-type Decoder struct {
-	rootRecord string
-	records    map[string]string
+// relay is the internal interface used to publish Pakrr messages to the DHT
+type relay interface {
+	Put(didID string, payload *bep44.Message) error
+	PutWithContext(ctx context.Context, didID string, payload *bep44.Message) error
+
+	Fetch(didID string) (*bep44.Message, error)
+	FetchWithContext(ctx context.Context, didID string) (*bep44.Message, error)
 }
 
-func (rec *Decoder) DIDDocument(didURI string) (*didcore.Document, error) {
-	if len(rec.rootRecord) == 0 {
-		return nil, errors.New("no root record found")
-	}
-	relationshipMap, err := parseVerificationRelationships(rec.rootRecord)
-	if err != nil {
-		return nil, err
-	}
+var defaultRelay relay
+var once sync.Once
 
-	// Now we have a did in a dns record. yay
-	document := &didcore.Document{
-		ID: didURI,
-	}
+// getDefaultRelay returns the default Pkarr relay client.
+func getDefaultRelay() relay {
+	once.Do(func() {
+		defaultRelay = pkarr.NewPkarrRelay("", http.DefaultClient)
+	})
 
-	// Now create the did document
-	for name, data := range rec.records {
-		switch {
-		case strings.HasPrefix(name, "_k"):
-			var vMethod didcore.VerificationMethod
-			if err := UnmarshalVerificationMethod(data, &vMethod); err != nil {
-				// TODO handle error
-				continue
-			}
+	return defaultRelay
+}
 
-			// extracting kN from _kN._did
-			entryID := strings.Split(name, ".")[0][1:]
-			relationships, ok := relationshipMap[entryID]
+// dHTDidOptions is used to configure the creation of a `did:dht` DID; it's an internal representation of the options
+type dHTDidOptions struct {
+	algorithmID         string
+	keyManager          crypto.KeyManager
+	services            []*didcore.Service
+	verificationMethods []didcore.VerificationMethod
+	purposes            map[string][]didcore.Purpose
+	relay               relay
+}
 
-			if !ok {
-				// no relationships
-				continue
-			}
+type DHTDidOption func(o *dHTDidOptions)
 
-			opts := []didcore.Purpose{}
-			for _, r := range relationships {
-				if o, ok := vmPurposeDNStoDID[r]; ok {
-					opts = append(opts, didcore.Purpose(o))
-				}
-			}
+// WithServices adds the provided services to the DID document.
+func WithServices(services ...*didcore.Service) DHTDidOption {
```

Comment:
i've left off `With` throughout the rest of the codebase thus far when using this options pattern. do you prefer including `With`? aiming to be consistent across the codebase.

Reply:
I'm indifferent so defer, but if we do stick with `With` convention, @mihai-chiorean can you open a backlog ticket to make consistent elsewhere?

Reply:
I think path of least resistance is just to match what we've done across the rest of the repo. Can open a GitHub issue if we want to change the naming pattern across the board and tackle later